packages/@aws-cdk/aws-certificatemanager/lib/dns_validated_certificate_handler/index.js

```diff
@@ -0,0 +1,196 @@
+'use strict'
+
+const aws = require('aws-sdk')
+const UuidEncoder = require('uuid-encoder')
+const request = require('request-promise-native')
+
+const sleep = function (ms) {
+  return new Promise(resolve => setTimeout(resolve, ms))
+}
+
+/**
+ * Requests a public certificate from AWS Certificate Manager, using DNS validation.
+ * The hosted zone ID must refer to a **public** Route53-managed DNS zone that is authoritative
+ * for the suffix of the certificate's Common Name (CN).  For example, if the CN is
+ * `*.example.com`, the hosted zone ID must point to a Route 53 zone authoritative
+ * for `example.com`.
+ *
+ * @param {string} requestId the CloudFormation request ID
+ * @param {string} domainName the Common Name (CN) field for the requested certificate
+ * @param {string} hostedZoneId the Route53 Hosted Zone ID
+ * @returns {string} Validated certificate ARN
+ */
+const requestCertificate = async function (requestId, domainName, hostedZoneId) {
+  const acm = new aws.ACM()
+  const route53 = new aws.Route53()
+  const encoder = new UuidEncoder('base36')
+
+  console.log(`Requesting certificate for ${domainName}`)
+
+  const reqCertResponse = await acm.requestCertificate({
+    DomainName: domainName,
+    IdempotencyToken: encoder.encode(requestId),
+    ValidationMethod: 'DNS'
+  }).promise()
+
+  console.log(`Certificate ARN: ${reqCertResponse.CertificateArn}`)
+
+  console.log('Waiting for ACM to provide DNS records for validation...')
+
+  var describeCertResponse
+  let attempt = 0
+  do {
+    // Exponential backoff based on 100ms base
+    await sleep(Math.pow(attempt, 2) * 100)
+    describeCertResponse = await acm.describeCertificate({
+      CertificateArn: reqCertResponse.CertificateArn
+    }).promise()
+  } while (describeCertResponse.Certificate.DomainValidationOptions[0].ValidationStatus !== 'PENDING_VALIDATION')
```

Comment:
Is this guaranteed not to end up in an error state?

This is making me feel uneasy. I'd rather check for the statuses you know you're waiting for (`=== 'PROVISIONING'` or whatever the status is) than this because you might miss some rare terminal state.

Reply:
Will address